### PR TITLE
Distill trainer

### PR DESCRIPTION
**Pull request by orenpereg**
1. Added Trainer for setfit distillation (file src/setfit/trainer_distill.py)
2. updated fewshot_distillation script to work with DistilSetFitTrainer

### CODE REVIEW

File: src/setfit/trainer_distill.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+        self._validate_column_mapping(self.train_dataset)
+        train_dataset = self.train_dataset
+        if self.column_mapping is not None:
+            logger.info("Applying column mapping to training dataset")
+            train_dataset = self._apply_column_mapping(self.train_dataset, self.column_mapping)
+
+        x_train = train_dataset["text"]
+        y_train = train_dataset["label"]
+        if self.loss_class is None:
```
Comment (lewtun) [inline src/setfit/trainer_distill.py:150]: Should we raise a proper `ValueError` here with a message or did I misunderstand the role of this `return` statement?

**Review by lewtun — COMMENTED** [review]
Thanks for adding this nice feature @orenpereg 🔥 !

I've left a few comments / questions and would love to see some unit tests in a `test_trainer_distillation.py` file (you can take inspiration from the existing `SetFitTrainer` ones)

Could you also share a code snippet of how this trainer works end-to-end? I know we have that logic in the script, but it would be great to have a condensed version we can later add to the README :)

File: scripts/setfit/run_fewshot_distillation.py [hunk 2]
```diff
@@ -245,18 +236,60 @@ def train(self):
...
+                    # self.model.load_state_dict(copy.deepcopy(self.model_wrapper.model_original_state))
+                    fewshot_ds[name].rename_column(
+                        "text", "sentence"
+                    )  # Map dataset columns to sentence/label expected by trainer
+                    teacher_model = SetFitModel.from_pretrained(self.model_name)
+                    teacher_trainer = SetFitTrainer(
+                        model=teacher_model,
+                        train_dataset=fewshot_ds[name],
+                        eval_dataset=eval_dataset,
+                        loss_class=CosineSimilarityLoss,
```
Comment (lewtun) [inline scripts/setfit/run_fewshot_distillation.py:249]: Should we add the type with something like this?

```suggestion
                        loss_class: torch.nn.Module =CosineSimilarityLoss,
```

File: scripts/setfit/run_fewshot_distillation.py [hunk 3]
```diff
@@ -10,22 +8,17 @@
...
 from xmlrpc.client import Boolean
 
-import numpy as np
 import pandas as pd
-from datasets import Dataset, DatasetDict, load_dataset
+from datasets import Dataset, DatasetDict, concatenate_datasets, load_dataset
 from distillation_baseline import BaselineDistillation
 from evaluate import load
+from sentence_transformers import losses
+from sentence_transformers.losses import CosineSimilarityLoss
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py:16]: thanks, done.

File: scripts/setfit/run_fewshot_distillation.py [hunk 4]
```diff
@@ -84,7 +77,7 @@ def parse_args():
 
 
 class RunFewShotDistill:
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py:79]: I ran the ag_news test with using 5 seeds. the results are very close. the diffs are between 0.1-0.6 points. i also checked emotion using a single seed and it looks quite similar. BTW the baseline model produces lower results. i still have to check why.
Screen Shot 2022-11-13 at 18 01 20

File: scripts/setfit/run_fewshot_distillation.py [hunk 5]
```diff
@@ -245,18 +236,60 @@ def train(self):
                 # if os.path.exists(results_path):
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py:236]: yes, done

File: scripts/setfit/run_fewshot_distillation.py [hunk 6]
```diff
@@ -245,18 +236,60 @@ def train(self):
...
-                if (self.mode == self.TEACHER) or (self.mode == self.SETFIT_STUDENT):
-                    self.model.load_state_dict(copy.deepcopy(self.model_wrapper.model_original_state))
-                    metrics = self.train_eval_setfit(
-                        fewshot_ds[name],
-                        test_dataset,
-                        self.loss_class,
-                        self.args.num_epochs,
-                        metric,
+                if self.mode == self.TEACHER:
+                    # self.model.load_state_dict(copy.deepcopy(self.model_wrapper.model_original_state))
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py:240]: yes, done.

File: scripts/setfit/run_fewshot_distillation.py [hunk 7]
```diff
@@ -271,83 +304,17 @@ def train_baseline_student(self, train_data, test_data, num_classes):
         x_test = test_data["text"]
         y_test = test_data["label"]
 
-        x_train = self.x_train_teacher + x_train
-        x_train_embd_student = self.trained_teacher_model.sbert_model.encode(x_train)
+        # x_train_embd_student = self.trained_teacher_model.sbert_model.encode(x_train)
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py:307]: yes, done :)

File: src/setfit/trainer_distill.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+import numpy as np
+import torch
+from sentence_transformers import InputExample, losses, util
+from sentence_transformers.datasets import SentenceLabelDataset
+from sentence_transformers.losses.BatchHardTripletLoss import BatchHardTripletLossDistanceFunction
+from torch.utils.data import DataLoader
+from transformers.trainer_utils import set_seed
+
+from setfit import SetFitTrainer
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply src/setfit/trainer_distill.py:12]: thanks! good point, done.

File: src/setfit/trainer_distill.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    from datasets import Dataset
+
+    from .modeling import SetFitModel
+
+logging.set_verbosity_info()
+logger = logging.get_logger(__name__)
+
+
+class DistilSetFitTrainer(SetFitTrainer):
+    """Trainer to train a distilled SetFit model (student) model.
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply src/setfit/trainer_distill.py:29]: yes, this sounds much better. done.

File: src/setfit/trainer_distill.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        use_amp (`bool`, *optional*, defaults to `False`):
+            Use Automatic Mixed Precision (AMP). Only for Pytorch >= 1.6.0
+        warmup_proportion (`float`, *optional*, defaults to `0.1`):
+            Proportion of the warmup in the total training steps.
+            Must be greater than or equal to 0.0 and less than or equal to 1.0.
+    """
+
+    def __init__(
+        self,
+        teacher_model: "SetFitModel" = None,
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply src/setfit/trainer_distill.py:70]: Right, thanks, i removed 'optional'

File: scripts/setfit/run_fewshot_distillation.py (hunk 2 above)
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py:249]: good point. done

File: src/setfit/trainer_distill.py (hunk 1 above)
Comment (orenpereg) [reply src/setfit/trainer_distill.py]: i agree, i added a ValueError with a message. BTW, we had the same issue in SetFitTrainer i also added a ValueError there.

File: scripts/setfit/run_fewshot_distillation.py [hunk 11]
```diff
@@ -245,18 +236,60 @@ def train(self):
...
+                    )  # Map dataset columns to sentence/label expected by trainer
+                    teacher_model = SetFitModel.from_pretrained(self.model_name)
+                    teacher_trainer = SetFitTrainer(
+                        model=teacher_model,
+                        train_dataset=fewshot_ds[name],
+                        eval_dataset=eval_dataset,
+                        loss_class=CosineSimilarityLoss,
+                        metric=metric,
+                        batch_size=self.args.batch_size,
+                        num_iterations=20,  # The number of text pairs to generate for contrastive learning
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py]: I agree. i changed to 'args.num_iterations_student' and 'args.num_iterations_teacher'.

File: scripts/setfit/run_fewshot_distillation.py [hunk 12]
```diff
@@ -245,18 +236,60 @@ def train(self):
...
+                    student_model = SetFitModel.from_pretrained(self.model_name)
+                    student_trainer = DistilSetFitTrainer(
+                        teacher_model=self.trained_teacher_model,
+                        model=student_model,
+                        train_dataset=student_train_dataset,
+                        eval_dataset=eval_dataset,
+                        loss_class=CosineSimilarityLoss,
+                        metric="accuracy",
+                        batch_size=16,
+                        num_iterations=20,  # The number of text pairs to generate for contrastive learning
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py]: yes. done :)

File: scripts/setfit/run_fewshot_distillation.py [hunk 13]
```diff
@@ -245,18 +236,60 @@ def train(self):
...
+
+                    self.treacher_train_data = fewshot_ds[name]  # save teacher training data
+                    self.trained_teacher_model = teacher_trainer.model
+
+                if self.mode == self.SETFIT_STUDENT:
+
+                    # student train data = teacher train data + unlabeled data
+                    student_train_dataset = concatenate_datasets([self.treacher_train_data, fewshot_ds[name]])
+                    student_train_dataset.rename_column(
+                        "text", "sentence"
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py:266]: You are right, I deleted it. the columns are normalized under the setfit org.

**Comment by orenpereg** [comment]
Hi @lewtun. Thanks for your thorough review and good comments🔥. I addressed all your comments and added tests under tests/test_trainer_distillation.py. can you pls take a look?
I also wrote a code snippet for the readme for using this trainer end-to-end:
[Run distillation using trainer example.docx](https://github.com/huggingface/setfit/files/10010292/Run.distillation.using.trainer.example.docx)

File: scripts/setfit/run_fewshot_distillation.py [hunk 14]
```diff
@@ -245,18 +236,60 @@ def train(self):
...
+                        teacher_model=self.trained_teacher_model,
+                        model=student_model,
+                        train_dataset=student_train_dataset,
+                        eval_dataset=eval_dataset,
+                        loss_class=CosineSimilarityLoss,
+                        metric="accuracy",
+                        batch_size=16,
+                        num_iterations=20,  # The number of text pairs to generate for contrastive learning
+                        num_epochs=1  # The number of epochs to use for constrastive learning
+                        # column_mapping={"sentence": "text", "label": "label"} # Map dataset columns to text/label expected by trainer
```
[the remark this answers is not in the document]
Comment (lewtun) [reply scripts/setfit/run_fewshot_distillation.py:274]: I think the unused code is still here :)

**Review by lewtun — COMMENTED** [review]
Thanks for iterating on this feature @orenpereg and adding some nice tests ❤️ !

I've left some final nits and then this should be good to merge :)

File: scripts/setfit/run_fewshot_distillation.py [hunk 15]
```diff
@@ -10,28 +8,21 @@
...
-from setfit.modeling import (
-    LOSS_NAME_TO_CLASS,
-    SetFitBaseModel,
-    SKLearnWrapper,
-    sentence_pairs_generation,
-    sentence_pairs_generation_cos_sim,
-)
+
+from setfit import DistillationSetFitTrainer, SetFitModel, SetFitTrainer
+from setfit.modeling import SetFitBaseModel, SKLearnWrapper
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py:19]: Thanks Lewis, I deleted it and also deleted the get_classifier function (that used SKLearnWrapper) which is not used

File: scripts/setfit/run_fewshot_distillation.py [hunk 16]
```diff
@@ -84,7 +78,7 @@ def parse_args():
 
 
 class RunFewShotDistill:
-    def __init__(self, args, mode, trained_teacher_model, x_train_teacher, student_train_ds) -> None:
+    def __init__(self, args, mode, trained_teacher_model, treacher_train_data, student_train_ds) -> None:
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply scripts/setfit/run_fewshot_distillation.py:81]: Thanks, good idea. done.

File: src/setfit/trainer.py [hunk 17]
```diff
@@ -287,7 +287,7 @@ def train(
         x_train = train_dataset["text"]
         y_train = train_dataset["label"]
         if self.loss_class is None:
-            return
+            raise RuntimeError("SetFitTrainer: training requires a loss_class argument")
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply src/setfit/trainer.py:290]: very good idea. this is much cleaner. i updated it in both Trainers

File: src/setfit/trainer_distillation.py [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+        teacher_model (`SetFitModel`):
+            The teacher model to mimic.
+        model (`SetFitModel`):
+            The student model to train. If not provided, a `model_init` must be passed.
+        train_dataset (`Dataset`):
+            The training dataset.
+        eval_dataset (`Dataset`, *optional*):
+            The evaluation dataset.
+        model_init (`Callable[[], SetFitModel]`, *optional*):
+            A function that instantiates the model to be used. If provided, each call to [`~DistilledSetFitTrainer:.train`] will start
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply src/setfit/trainer_distillation.py:39]: done

File: src/setfit/trainer_distillation.py [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+        use_amp (`bool`, *optional*, defaults to `False`):
+            Use Automatic Mixed Precision (AMP). Only for Pytorch >= 1.6.0
+        warmup_proportion (`float`, *optional*, defaults to `0.1`):
+            Proportion of the warmup in the total training steps.
+            Must be greater than or equal to 0.0 and less than or equal to 1.0.
+    """
+
+    def __init__(
+        self,
+        teacher_model: "SetFitModel" = None,
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply src/setfit/trainer_distillation.py:68]: right. done

File: src/setfit/trainer_distillation.py [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+            l2_weight (float, *optional*):
+                Temporary change the weight of L2 regularization for SetFitModel's differentiable head in logistic regression.
+            trial (`optuna.Trial` or `Dict[str, Any]`, *optional*):
+                The trial run or the hyperparameter dictionary for hyperparameter search.
+        """
+        if trial:  # Trial and model initialization
+            set_seed(self.seed)  # Seed must be set before instantiating the model when using model_init.
+            self._hp_search_setup(trial)  # sets trainer parameters and initializes model
+
+        if self.teacher_model is None:
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply src/setfit/trainer_distillation.py:137]: yes. done.

File: src/setfit/trainer_distillation.py [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+            raise RuntimeError("`DistillationSetFitTrainer` training requires a teacher model")
+
+        if self.train_dataset is None:
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply src/setfit/trainer_distillation.py:140]: done

File: tests/test_trainer_distillation.py [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+            eval_dataset=dataset,
+            num_iterations=self.num_iterations,
+            column_mapping={"text_new": "text", "label_new": "label"},
+        )
+
+        trainer._validate_column_mapping(trainer.train_dataset)
+        formatted_dataset = trainer._apply_column_mapping(trainer.train_dataset, trainer.column_mapping)
+
+        assert formatted_dataset.column_names == ["text", "label"]
+
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply tests/test_trainer_distillation.py:108]: done

File: tests/test_trainer_distillation.py [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+        assert formatted_dataset[0]["text"] == "a"
+        assert formatted_dataset[0]["label"] == [0, 1]
+
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply tests/test_trainer_distillation.py:111]: done

File: src/setfit/trainer_distillation.py [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+
+class DistillationSetFitTrainer(SetFitTrainer):
+    """Trainer to compress a SetFit model with knowledge distillation.
+
+    Args:
+        teacher_model (`SetFitModel`):
+            The teacher model to mimic.
+        model (`SetFitModel`):
+            The student model to train. If not provided, a `model_init` must be passed.
+        train_dataset (`Dataset`):
```
[the remark this answers is not in the document]
Comment (orenpereg) [reply src/setfit/trainer_distillation.py:32]: thanks! i changes the order of the args and renamed model to student_model

**Review by lewtun — APPROVED** [review]
Thanks for the last round of changes @orenpereg - this looks really nice 🔥

**Comment by orenpereg** [comment]
thanks Lewis :)
